File: problems/simon-ionization-excess/compute/q14/verify_span.py
#!/usr/bin/env python3
"""Stdlib-only independent reconstruction of the q14 span bound.

Uses Decimal rather than mpmath intervals, parses the frozen q13 matrix
and face summary directly, and adds a 1e-40 outward safety pad.
"""

from __future__ import annotations

import json
from decimal import Decimal, getcontext
from pathlib import Path

HERE = Path(__file__).resolve().parent
CERTS = HERE / "certs"
Q13 = HERE.parent / "q13" / "certs"
MATRIX = Q13 / "beta3_mid_R10_n37_t0p9119.txt"
FACES = Q13 / "beta3_mid_faces_R10_n37_t0p9119.txt"

getcontext().prec = 80
D = Decimal
PAD = D("1e-40")
MATRIX_PAD = D("1e-12")


def f(t: Decimal) -> Decimal:
    return (1 + t**3) / (1 + t**2)


def face_map() -> dict[str, str]:
    ans = {}
    for line in FACES.read_text().splitlines():
        key, _, value = line.partition(" ")
        ans[key] = value.strip()
    return ans


def main() -> None:
    R = D(10)
    n = 37
    target = D("0.9119")
    q = (R.ln() / D(n)).exp()
    P_hi = (q - 1) / (q + 1) + PAD
    u = ((1 + D(2).sqrt()).ln() / D(3)).exp()
    t_star = u - 1 / u
    f_min_lo = D(3) * t_star / D(2) - PAD
    t_far = q * q / R
    t_near = 1 / q
    f_far_hi = f(t_far) + PAD
    f_near_hi = f(t_near) + PAD
    span_hi = max(f_far_hi, f_near_hi) - f_min_lo
    error_hi = P_hi * span_hi + MATRIX_PAD
    gamma_lo = target - error_hi
    leading_hi = 1 / gamma_lo + PAD
    cut = R / (R + 1)

    faces = face_map()
    nfaces = (1 << n) - 1
    face_ok = (
        int(faces["n"]) == n
        and abs(D(faces["gamma_target"]) - target) < D("1e-15")
        and int(faces["n_faces"]) == nfaces
        and int(faces["gray_i"]) == nfaces
        and int(faces["copositive"]) == 1
        and int(faces["singular_or_illconditioned"]) == 0
        and D(faces["min_mMm_safe"]) > 0
    )

    toks = MATRIX.read_text().split()
    matrix_n = int(toks[0])
    matrix_target = D(toks[1])
    c = [D(x) for x in toks[2 : 2 + matrix_n]]
    flat_A = [D(x) for x in toks[2 + matrix_n :]]
    if len(flat_A) != matrix_n * matrix_n:
        raise SystemExit("bad matrix length")
    stored_F = []
    for i in range(matrix_n):
        for j in range(matrix_n):
            aij = flat_A[i * matrix_n + j]
            stored_F.append(2 * aij / (c[i] + c[j]))
    stored_min = min(stored_F)
    stored_max = max(stored_F)

    checks = {
        "face_certificate_ok": face_ok,
        "matrix_header_ok": matrix_n == n and abs(matrix_target - target) < D("1e-15"),
        "far_endpoint_below_t_star": t_far < t_star,
        "near_endpoint_above_t_star": t_near > t_star,
        "far_value_dominates_near": f_far_hi > f_near_hi,
        "stored_F_above_f_min": stored_min + D("1e-14") >= f_min_lo,
        "stored_F_below_analytic_max": stored_max <= f_far_hi + D("1e-14"),
        "gamma_below_cut": gamma_lo < cut,
        "leading_below_1.1002": leading_hi < D("1.1002"),
        "leading_below_q13_1.1006": leading_hi < D("1.1006"),
    }
    ok = all(checks.values())
    blob = {
        "implementation": "Python stdlib Decimal, 80 digits, 1e-40 transcendental pad",
        "matrix_rounding_pad": str(MATRIX_PAD),
        "q": str(q),
        "P_upper": str(P_hi),
        "t_star": str(t_star),
        "f_min_lower": str(f_min_lo),
        "f_far_upper": str(f_far_hi),
        "f_near_upper": str(f_near_hi),
        "stored_F_min": str(stored_min),
        "stored_F_max": str(stored_max),
        "span_upper": str(span_hi),
        "error_upper": str(error_hi),
        "beta3_lower": str(gamma_lo),
        "leading_upper": str(leading_hi),
        "cut": str(cut),
        "checks": checks,
        "ok": ok,
    }
    out = CERTS / "span_stdlib.json"
    out.write_text(json.dumps(blob, indent=2) + "\n")
    print("stored F range", stored_min, stored_max)
    print("span <=", span_hi, "error <=", error_hi)
    print("beta3 >=", gamma_lo, "leading <=", leading_hi)
    print("wrote", out)
    if not ok:
        raise SystemExit("verify_span.py FAIL")
    print("verify_span.py PASS")


if __name__ == "__main__":
    main()

File: problems/simon-ionization-excess/compute/q14/verify_span.rs
//! Independent Rust replay of the q14 finite-range tax.
//!
//! Reads the frozen q13 matrix and Gray summary, reconstructs the stored
//! F range, and checks the analytic geometric-bin envelope with explicit
//! f64 safety pads.  No Python modules or shared helpers.

use std::fs;

fn kernel(t: f64) -> f64 {
    (1.0 + t * t * t) / (1.0 + t * t)
}

fn face_value(text: &str, wanted: &str) -> String {
    for line in text.lines() {
        let mut fields = line.split_whitespace();
        if fields.next() == Some(wanted) {
            return fields.next().expect("missing face value").to_string();
        }
    }
    panic!("missing face key {}", wanted);
}

fn main() {
    let base = "../q13/certs";
    let matrix_path = format!("{base}/beta3_mid_R10_n37_t0p9119.txt");
    let faces_path = format!("{base}/beta3_mid_faces_R10_n37_t0p9119.txt");
    let matrix = fs::read_to_string(&matrix_path).expect("read q13 matrix");
    let faces = fs::read_to_string(&faces_path).expect("read q13 faces");
    let mut tok = matrix.split_whitespace();
    let n: usize = tok.next().unwrap().parse().unwrap();
    let target: f64 = tok.next().unwrap().parse().unwrap();
    assert_eq!(n, 37);
    assert!((target - 0.9119).abs() < 1e-15);
    let mut c = Vec::with_capacity(n);
    for _ in 0..n {
        c.push(tok.next().unwrap().parse::<f64>().unwrap());
    }
    let mut a = Vec::with_capacity(n * n);
    for _ in 0..(n * n) {
        a.push(tok.next().unwrap().parse::<f64>().unwrap());
    }
    assert!(tok.next().is_none());

    let mut stored_min = f64::INFINITY;
    let mut stored_max = f64::NEG_INFINITY;
    for i in 0..n {
        for j in 0..n {
            let fij = 2.0 * a[i * n + j] / (c[i] + c[j]);
            stored_min = stored_min.min(fij);
            stored_max = stored_max.max(fij);
        }
    }

    let total = (1u64 << n) - 1;
    let face_ok = face_value(&faces, "n").parse::<usize>().unwrap() == n
        && (face_value(&faces, "gamma_target").parse::<f64>().unwrap() - target).abs() < 1e-15
        && face_value(&faces, "n_faces").parse::<u64>().unwrap() == total
        && face_value(&faces, "gray_i").parse::<u64>().unwrap() == total
        && face_value(&faces, "copositive").parse::<u32>().unwrap() == 1
        && face_value(&faces, "singular_or_illconditioned").parse::<u64>().unwrap() == 0
        && face_value(&faces, "min_mMm_safe").parse::<f64>().unwrap() > 0.0;
    assert!(face_ok);

    let r = 10.0_f64;
    let q = r.powf(1.0 / n as f64);
    let p_hi = (q - 1.0) / (q + 1.0) + 1e-14;
    let u = (1.0 + 2.0_f64.sqrt()).powf(1.0 / 3.0);
    let t_star = u - 1.0 / u;
    let f_min_lo = 1.5 * t_star - 1e-14;
    let t_far = q * q / r;
    let t_near = 1.0 / q;
    let f_far_hi = kernel(t_far) + 1e-14;
    let f_near_hi = kernel(t_near) + 1e-14;
    assert!(t_far < t_star && t_star < t_near);
    assert!(f_near_hi < f_far_hi);
    assert!(stored_min + 1e-12 >= f_min_lo);
    assert!(stored_max <= f_far_hi + 1e-12);

    let span_hi = f_far_hi - f_min_lo;
    let matrix_pad = 1e-12;
    let error_hi = p_hi * span_hi + matrix_pad;
    let gamma_lo = target - error_hi;
    let leading_hi = 1.0 / gamma_lo + 1e-14;
    let cut = 10.0 / 11.0;
    assert!(gamma_lo < cut);
    assert!(leading_hi < 1.1002);
    assert!(leading_hi < 1.1006);

    let out = format!(
        concat!(
            "{{\n",
            "  \"implementation\": \"Rust f64 with 1e-14 outward pads\",\n",
            "  \"stored_F_min\": {:.17e},\n",
            "  \"stored_F_max\": {:.17e},\n",
            "  \"analytic_F_max_upper\": {:.17e},\n",
            "  \"span_upper\": {:.17e},\n",
            "  \"error_upper\": {:.17e},\n",
            "  \"matrix_rounding_pad\": {:.17e},\n",
            "  \"beta3_lower\": {:.17e},\n",
            "  \"leading_upper\": {:.17e},\n",
            "  \"cut\": {:.17e},\n",
            "  \"face_certificate_ok\": true,\n",
            "  \"ok\": true\n",
            "}}\n"
        ),
        stored_min, stored_max, f_far_hi, span_hi, error_hi, matrix_pad,
        gamma_lo, leading_hi, cut
    );
    fs::create_dir_all("certs").unwrap();
    fs::write("certs/span_rs.json", out).unwrap();
    println!("stored F range [{stored_min:.12}, {stored_max:.12}]");
    println!("beta3 >= {gamma_lo:.15}, leading <= {leading_hi:.15}");
    println!("verify_span.rs PASS");
}
